### PR TITLE
feat(broker): snapshot director can distinguish processing and replay mode

### LINKED ISSUE #7510: Snapshots on Followers
**Description**

This is part of the [ZEP 0007](https://github.com/zeebe-io/enhancements/blob/master/ZEP007-build-state-on-followers.md)

After we build state on followers with the ReplayStateMachine, we need to take snapshots from it to be able to compact our log. 

From the ZEP:
> In order to take valid snapshots on the Leader we need to wait until the last written position of the stream processor is committed. This is necessary, because when we process a command we produce follow-up events, which correspond to our state changes. These state changes are immediately applied during processing, which means they will be part of the snapshot. In order to restore the same state on restart and not process or apply events twice we have to wait until the events, which we have written are committed.
>
> On Follower's it is a bit different. The Follower applies events only. The last processed position here corresponds to the source position of the last applied event, which means the last written position corresponds here to the last applied event position. In general the follower doesn't need to wait until the last written position is committed, since he can only read committed entries anyway.

As part of this issue we could either create a simpler version of the snapshot director or parameterize the existing one, such that it doesn't wait until the last written position is committed.

Blocked by https://github.com/camunda-cloud/zeebe/issues/7511

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

On replay mode, snapshot director does not wait for the the last event to commit because only committed events are replayed. Hence all the events included in the snapshot is already committed. So it is safe for the snapshot director to immediately persist a snapshot.

When running on replay mode (on the follower) the commit listeners are not available. So it is not possible to wait for the commit. This is why we need different logic for snapshot director running on the follower and the leader.

## Related issues

closes #7510 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshotingOnFollowerTest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+  @Test
+  public void shouldPersistSnapshotWithoutWaitingForCommit() {
+    // when
+    asyncSnapshotDirector.forceSnapshot();
+
+    // then
+    waitUntil(() -> snapshotController.getValidSnapshotsCount() == 1);
+    assertThat(snapshotController.getValidSnapshotsCount()).isEqualTo(1);
+  }
```
Comment (saig0) [inline broker/src/test/java/io/camunda/zeebe/broker/system/partitions/AsyncSnapshotingOnFollowerTest.java:119]: Can we merge the test case to `AsyncSnapshotingTest`?

The setup code looks very similar to the existing test. I'm not sure if it is worth copying so much code for one test case.

**Review by saig0 — COMMENTED** [review]
Looks good :+1: 

Please see my comment about the test case.

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java [hunk 2]
```diff
@@ -55,17 +57,65 @@
...
 
-  public AsyncSnapshotDirector(
+  private AsyncSnapshotDirector(
       final int nodeId,
       final int partitionId,
       final StreamProcessor streamProcessor,
       final StateController stateController,
-      final Duration snapshotRate) {
+      final Duration snapshotRate,
+      final boolean replayMode) {
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java:66]: Using ReplayMode is a bit confusing, and has no meaning outside of ReplayStateMachine. Any chance we can rename it to ProcessingMode?
ReplayMode.UNTIL_END -> ProcessingMode.PROCESSING
ReplayMode.CONTINUOUS -> ProcessingMode.REPLAY

[the remark this answers is not in the document]
Comment (saig0) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java:66]: Good point :thinking: 

I'm fine with the two values. But I'm not sure about the name `ProcessingMode`. What about `StreamProcessorMode`?

[the remark this answers is not in the document]
Comment (deepthidevaki) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/AsyncSnapshotDirector.java:66]: `StreamProcessorMode` is better :+1:

I will rename it in a follow up issue.

**Comment by deepthidevaki** [comment]
@saig0 I have applied your comments. I ended up refactoring `AsyncSnapshotingTest` a little bit. :slightly_smiling_face: Please take a look.

**Review by saig0 — APPROVED** [review]
Looks good :+1: Thanks :sunflower:

**Pull request merged**